gr-digital/lib/ofdm_cyclic_prefixer_impl.cc

```diff
@@ -31,135 +31,147 @@ namespace gr {
 
     ofdm_cyclic_prefixer::sptr
-    ofdm_cyclic_prefixer::make(size_t input_size, size_t output_size, int rolloff_len, const std::string &len_tag_key)
+    ofdm_cyclic_prefixer::make(int fft_len, std::vector<int> cp_lengths, int rolloff_len, const std::string &len_tag_key)
     {
-      return gnuradio::get_initial_sptr (new ofdm_cyclic_prefixer_impl(input_size, output_size, rolloff_len, len_tag_key));
+      return gnuradio::get_initial_sptr(new ofdm_cyclic_prefixer_impl(fft_len, cp_lengths, rolloff_len, len_tag_key));
     }
 
-
-    ofdm_cyclic_prefixer_impl::ofdm_cyclic_prefixer_impl(size_t input_size, size_t output_size, int rolloff_len, const std::string &len_tag_key)
-  : tagged_stream_block ("ofdm_cyclic_prefixer",
-		  io_signature::make (1, 1, input_size*sizeof(gr_complex)),
-		  io_signature::make (1, 1, sizeof(gr_complex)),
-		  len_tag_key),
-	d_fft_len(input_size),
-	d_output_size(output_size),
-	d_cp_size(output_size - input_size),
-	d_rolloff_len(rolloff_len),
-	d_up_flank((rolloff_len ? rolloff_len-1 : 0), 0),
-	d_down_flank((rolloff_len ? rolloff_len-1 : 0), 0),
-	d_delay_line(0, 0)
+    ofdm_cyclic_prefixer_impl::ofdm_cyclic_prefixer_impl(int fft_len, std::vector<int> cp_lengths, int rolloff_len, const std::string &len_tag_key)
+      : gr::tagged_stream_block
+        (
+          "ofdm_cyclic_prefixer",
+          gr::io_signature::make(1, 1, fft_len*sizeof(gr_complex)),
+          gr::io_signature::make(1, 1, sizeof(gr_complex)), len_tag_key
+        ),
+        d_fft_len(fft_len),
+        d_state(0),
+        d_cp_max(0),
+        d_rolloff_len(rolloff_len),
+        d_cp_lengths(cp_lengths),
+        d_up_flank((rolloff_len ? rolloff_len-1 : 0), 0),
+        d_down_flank((rolloff_len ? rolloff_len-1 : 0), 0),
+        d_delay_line(0, 0),
+        d_len_tag_key(len_tag_key)
     {
-      set_relative_rate(d_output_size);
-
-      // Flank of length 1 would just be rectangular
-      if (d_rolloff_len == 1) {
-	d_rolloff_len = 0;
+      for(const int& cp_len : d_cp_lengths) {
+        d_cp_max = std::max(d_cp_max, cp_len);
       }
-      if (d_rolloff_len) {
-	d_delay_line.resize(d_rolloff_len-1, 0);
-	if (rolloff_len > d_cp_size) {
-	  throw std::invalid_argument("cyclic prefixer: rolloff len must smaller than the cyclic prefix.");
-	}
-	// The actual flanks are one sample shorter than d_rolloff_len, because the
-	// first sample of the up- and down flank is always zero and one, respectively
-	for (int i = 1; i < d_rolloff_len; i++) {
-	  d_up_flank[i-1]   = 0.5 * (1 + cos(M_PI * i/rolloff_len - M_PI));
-	  d_down_flank[i-1] = 0.5 * (1 + cos(M_PI * (rolloff_len-i)/rolloff_len - M_PI));
-	}
+      // Sanity
+      for(unsigned i=0; i<d_cp_lengths.size(); i++) {
+        if (d_cp_lengths[i] != 0) {
+          break;
+        }
+        if (i == d_cp_lengths.size()-1) {
+          throw std::invalid_argument(this->alias() + std::string(": It does not make sense to only have one or a couple of CPs with length 0."));
```

Comment:
Prefer a positive error message ("You must provide X") instead of negative ("What you provided is invalid") so it's clear, based on the error message, what the user has to change.